### PR TITLE
[modernize] range-based loops 

### PR DESCRIPTION
**Pull request by Rechi**
## Description
run `run-clang-tidy -header-filter=.* -checks=-*,modernize-loop-convert -fix -- xbmc` for iOS, Linux (gbm, Wayland, X11) and OSX with the following configuration
```
CheckOptions:
	- key: 'modernize-loop-convert.MaxCopySize'
	  value: '8'
	- key: 'modernize-loop-convert.MinConfidence'
	  value: 'safe'
```

## How Has This Been Tested?
compiled for Android, iOS, Linux (gbm, Wayland, X11), OSX and Windows (Desktop, UWP)

## Types of change
- [ ] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

## Checklist:
- [x] My code follows the **[Code Guidelines](https://github.com/xbmc/xbmc/blob/master/docs/CODE_GUIDELINES.md)** of this project 
- [ ] My change requires a change to the documentation, either Doxygen or wiki
- [ ] I have updated the documentation accordingly
- [x] I have read the **[Contributing](https://github.com/xbmc/xbmc/blob/master/docs/CONTRIBUTING.md)** document
- [ ] I have added tests to cover my change
- [x] All new and existing tests passed

### CODE REVIEW

File: xbmc/cores/AudioEngine/Utils/AEChannelInfo.cpp [hunk 1]
```diff
@@ -121,8 +121,8 @@ void CAEChannelInfo::ResolveChannels(const CAEChannelInfo& rhs)
 void CAEChannelInfo::Reset()
 {
   m_channelCount = 0;
-  for (unsigned int i = 0; i < AE_CH_MAX; ++i)
-    m_channels[i] = AE_CH_NULL;
+  for (AEChannel& channel : m_channels)
```
Comment (notspiff) [inline xbmc/cores/AudioEngine/Utils/AEChannelInfo.cpp]: std::fill ?

File: xbmc/dialogs/GUIDialogMediaFilter.cpp [hunk 2]
```diff
@@ -393,13 +393,11 @@ void CGUIDialogMediaFilter::InitializeSettings()
     return;
   }
 
-  for (unsigned int index = 0; index < NUM_FILTERS; index++)
+  for (Filter filter : filterList)
```
Comment (notspiff) [inline xbmc/dialogs/GUIDialogMediaFilter.cpp]: NUM_FILTERS macro now unused.

File: xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp [hunk 3]
```diff
@@ -503,27 +503,27 @@ void CGUIDialogSmartPlaylistEditor::OnDeinitWindow(int nextWindowID)
 
 CGUIDialogSmartPlaylistEditor::PLAYLIST_TYPE CGUIDialogSmartPlaylistEditor::ConvertType(const std::string &type)
 {
-  for (unsigned int i = 0; i < NUM_TYPES; i++)
-    if (type == types[i].string)
-      return types[i].type;
+  for (const translateType& t : types)
```
Comment (notspiff) [inline xbmc/dialogs/GUIDialogSmartPlaylistEditor.cpp]: NUM_TYPES macro now unused.

File: xbmc/playlists/SmartPlayList.cpp [hunk 4]
```diff
@@ -148,15 +148,15 @@ CSmartPlaylistRule::CSmartPlaylistRule() = default;
 
 int CSmartPlaylistRule::TranslateField(const char *field) const
 {
-  for (unsigned int i = 0; i < NUM_FIELDS; i++)
```
Comment (notspiff) [inline xbmc/playlists/SmartPlayList.cpp]: NUM_FIELDS now unused.

File: xbmc/playlists/SmartPlayList.cpp [hunk 5]
```diff
@@ -176,44 +176,44 @@ std::string CSmartPlaylistRule::TranslateOrder(SortBy order)
 
 Field CSmartPlaylistRule::TranslateGroup(const char *group)
 {
-  for (unsigned int i = 0; i < NUM_GROUPS; i++)
```
Comment (notspiff) [inline xbmc/playlists/SmartPlayList.cpp]: NUM_GROUPS now unused.

File: xbmc/music/Album.cpp [hunk 6]
```diff
@@ -421,9 +421,8 @@ void CAlbum::SetLastPlayed(const std::string& strLastPlayed)
 
 std::string CAlbum::ReleaseTypeToString(CAlbum::ReleaseType releaseType)
 {
-  for (size_t i = 0; i < RELEASE_TYPES_SIZE; i++)
```
Comment (notspiff) [inline xbmc/music/Album.cpp]: RELEASE_TYPES_SIZE now unused.

File: xbmc/addons/AddonInfo.cpp [hunk 7]
```diff
@@ -72,39 +72,36 @@ static const TypeMapping types[] =
 
 std::string CAddonInfo::TranslateType(ADDON::TYPE type, bool pretty/*=false*/)
 {
-  for (unsigned int index=0; index < ARRAY_SIZE(types); ++index)
+  for (const TypeMapping& t : types)
```
Comment (notspiff) [inline xbmc/addons/AddonInfo.cpp]: call this map and diff shrinks.

File: xbmc/addons/AddonInfo.cpp [hunk 8]
```diff
@@ -72,39 +72,36 @@ static const TypeMapping types[] =
...
+        return t.name;
     }
   }
   return "";
 }
 
 TYPE CAddonInfo::TranslateType(const std::string& string)
 {
-  for (unsigned int index=0; index < ARRAY_SIZE(types); ++index)
+  for (const TypeMapping& t : types)
```
Comment (notspiff) [inline xbmc/addons/AddonInfo.cpp]: map

File: xbmc/addons/AddonInfo.cpp [hunk 9]
```diff
@@ -72,39 +72,36 @@ static const TypeMapping types[] =
...
+      return t.type;
   }
 
   return ADDON_UNKNOWN;
 }
 
 std::string CAddonInfo::TranslateIconType(ADDON::TYPE type)
 {
-  for (unsigned int index = 0; index < ARRAY_SIZE(types); ++index)
+  for (const TypeMapping& t : types)
```
Comment (notspiff) [inline xbmc/addons/AddonInfo.cpp]: map

File: xbmc/addons/Scraper.cpp [hunk 10]
```diff
@@ -60,27 +60,25 @@ static const ContentMapping content[] = {{"unknown", CONTENT_NONE, 231},
 
 std::string TranslateContent(const CONTENT_TYPE &type, bool pretty /*=false*/)
 {
-  for (unsigned int index = 0; index < ARRAY_SIZE(content); ++index)
+  for (const ContentMapping& contentMapping : content)
```
Comment (notspiff) [inline xbmc/addons/Scraper.cpp]: map

File: xbmc/addons/Scraper.cpp [hunk 11]
```diff
@@ -60,27 +60,25 @@ static const ContentMapping content[] = {{"unknown", CONTENT_NONE, 231},
...
+        return contentMapping.name;
     }
   }
   return "";
 }
 
 CONTENT_TYPE TranslateContent(const std::string &string)
 {
-  for (unsigned int index = 0; index < ARRAY_SIZE(content); ++index)
+  for (const ContentMapping& contentMapping : content)
```
Comment (notspiff) [inline xbmc/addons/Scraper.cpp]: map

File: xbmc/cores/AudioEngine/Sinks/AESinkSNDIO.cpp [hunk 12]
```diff
@@ -88,15 +88,15 @@ void CAESinkSNDIO::AudioFormatToPar(AEAudioFormat& format)
   m_par.xrun = SIO_IGNORE;
   m_par.pchan = format.m_channelLayout.Count();
 
-  for (size_t i = 0; i < nitems(formats); i++)
+  for (const sndio_formats& f : formats)
```
Comment (FernetMenta) [inline xbmc/cores/AudioEngine/Sinks/AESinkSNDIO.cpp]: above it's format and here just f

File: xbmc/LangInfo.cpp [hunk 13]
```diff
@@ -152,9 +152,8 @@ static CTemperature::Unit StringToTemperatureUnit(const std::string& temperature
   std::string unit(temperatureUnit);
   StringUtils::ToLower(unit);
 
-  for (size_t i = 0; i < TEMPERATURE_INFO_SIZE; i++)
+  for (const TemperatureInfo& info : temperatureInfo)
```
Comment (notspiff) [inline xbmc/LangInfo.cpp]: TEMPERATURE_INFO_SIZE now unused.

File: xbmc/LangInfo.cpp [hunk 14]
```diff
@@ -167,9 +166,8 @@ static CSpeed::Unit StringToSpeedUnit(const std::string& speedUnit)
   std::string unit(speedUnit);
   StringUtils::ToLower(unit);
 
-  for (size_t i = 0; i < SPEED_INFO_SIZE; i++)
```
Comment (notspiff) [inline xbmc/LangInfo.cpp]: SPEED_INFO_SIZE now unused

File: xbmc/LangInfo.cpp [hunk 15]
```diff
@@ -1177,9 +1175,8 @@ void CLangInfo::SettingOptionsShortDateFormatsFiller(SettingConstPtr setting, st
     current = SETTING_REGIONAL_DEFAULT;
   }
 
-  for (size_t i = 0; i < SHORT_DATE_FORMATS_SIZE; i++)
```
Comment (notspiff) [inline xbmc/LangInfo.cpp]: SHORT_DATE_FORMATS_SIZE now unused

File: xbmc/LangInfo.cpp [hunk 16]
```diff
@@ -1207,9 +1204,8 @@ void CLangInfo::SettingOptionsLongDateFormatsFiller(SettingConstPtr setting, std
     current = SETTING_REGIONAL_DEFAULT;
   }
 
-  for (size_t i = 0; i < LONG_DATE_FORMATS_SIZE; i++)
```
Comment (notspiff) [inline xbmc/LangInfo.cpp]: LONG_DATE_FORMATS_SIZE now unused

File: xbmc/cores/AudioEngine/Sinks/AESinkSNDIO.cpp (hunk 12 above)
Comment (Rechi) [reply xbmc/cores/AudioEngine/Sinks/AESinkSNDIO.cpp]: `format` variable name is already taken by a function parameter. If you want I can adjust the above name.

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp [hunk 17]
```diff
@@ -172,21 +172,21 @@ int compWindowsPriorities (const void *a, const void *b)
...
+    c[I708_SCREENGRID_COLUMNS]=0;
   }
 };
 
 void printTVtoBuf (cc708_service_decoder *decoder)
 {
   int empty=1;
   decoder->textlen = 0;
-  for (int i=0;i<75;i++)
+  for (auto& c : decoder->tv.chars)
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp]: 75 is not entire range

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp [hunk 18]
```diff
@@ -197,23 +197,23 @@ void printTVtoBuf (cc708_service_decoder *decoder)
   if (empty)
     return; // Nothing to write
 
-  for (int i=0;i<75;i++)
+  for (auto& c : decoder->tv.chars)
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp]: we iterate over rows

**Review by notspiff — CHANGES_REQUESTED** [review]

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp [hunk 19]
```diff
@@ -172,21 +172,21 @@ int compWindowsPriorities (const void *a, const void *b)
 
 void clearTV (cc708_service_decoder *decoder)
 {
-  for (int i=0; i<I708_SCREENGRID_ROWS; i++)
+  for (auto& c : decoder->tv.chars)
```
[the remark this answers is not in the document]
Comment (Rechi) [reply xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp]: `auto& c` is deduced to `unsigned char (c&)[I708_SCREENGRID_COLUMNS+1]`. If wanted I can make it explicit.

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp (hunk 18 above)
Comment (Rechi) [reply xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp]: still happening, `auto& c` is deduced to `unsigned char (c&)[I708_SCREENGRID_COLUMNS+1]`

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp (hunk 17 above)
Comment (Rechi) [reply xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp]: `unsigned char chars[I708_SCREENGRID_ROWS][I708_SCREENGRID_COLUMNS+1];` and `#define I708_SCREENGRID_ROWS 75` so why should 75 not be the entire range?

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp [hunk 20]
```diff
@@ -2056,15 +2055,15 @@ void CLinuxRendererGL::DeleteYV12Texture(int index)
...
     {
-      if (m_buffers[index].fields[f][p].id)
+      if (p.id)
       {
-        if (glIsTexture(m_buffers[index].fields[f][p].id))
-          glDeleteTextures(1, &m_buffers[index].fields[f][p].id);
-        m_buffers[index].fields[f][p].id = 0;
+        if (glIsTexture(p.id))
+          glDeleteTextures(1, &p.id);
+        p.id = 0;
```
Comment (FernetMenta) [inline xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp]: it is important to be aware of field and plane index. there are still places left where range based loops won't fit, i.e.

```
  for (int f = FIELD_FULL; f<=FIELD_BOT ; f++)
  {
    int fieldshift = (f==FIELD_FULL) ? 0 : 1;
```

mixing two concepts make the code harder to read

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp]: it should be clear that we iterate the rows here.

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp (hunk 17 above)
Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp]: see above, not clear enough what we get from auto c

File: xbmc/LangInfo.cpp (hunk 13 above)
Comment (Rechi) [reply xbmc/LangInfo.cpp]: stupid defines don't generate a unused-something warning

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp (hunk 20 above)
Comment (Rechi) [reply xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp]: Where should I revert it back to the current style, only here, the whole file or additionally also `LinuxRendererGLES.cpp`?

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (Rechi) [reply xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp]: replaced `auto& c` with `unsigned char(&row)[I708_SCREENGRID_COLUMNS + 1]`

File: xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp (hunk 20 above)
Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/VideoRenderers/LinuxRendererGL.cpp]: VideoRenderers/* seems easiest to me

File: xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp (hunk 19 above)
[the remark this answers is not in the document]
Comment (FernetMenta) [reply xbmc/cores/VideoPlayer/DVDCodecs/Overlay/contrib/cc_decoder708.cpp]: great, thanks

File: xbmc/cores/AudioEngine/Utils/AEChannelInfo.cpp (hunk 1 above)
Comment (Rechi) [reply xbmc/cores/AudioEngine/Utils/AEChannelInfo.cpp]: Not sure if  I like, `std::fill(m_channels, m_channels + sizeof(m_channels) / sizeof(AEChannel), AE_CH_NULL);` (m_channels is a C array).
What do you think?

Comment (notspiff) [reply xbmc/cores/AudioEngine/Utils/AEChannelInfo.cpp]: less readable, keep the for.

**Review by notspiff — APPROVED** [review]

**Review by FernetMenta — APPROVED** [review]

**Comment by HitcherUK** [comment]
@Rechi This seems to be the cause of slowing down the loading of media windows (TV Shows/Movies).

https://forum.kodi.tv/showthread.php?tid=335629

Logs show nothing except delays after loading -

https://forum.kodi.tv/showthread.php?tid=298462&pid=2768566#pid2768566

I was seeing the same behaviour when entering TV Shows with 11 Sept build but not with 10th Sept.

**Comment by Rechi** [comment]
> https://forum.kodi.tv/showthread.php?tid=298462&pid=2768566#pid2768566

Measuring time for different directories of different users doesn't say anything.

**Comment by fritsch** [comment]
Right, but his last sentence made me curious: that was obviously the "same TV shows".

**Comment by HitcherUK** [comment]
So what will?

As I said nothing is shown in the debug logs except the time difference between the 2 builds.

10 Sept -

`13:03:02.161 T:138296648944   DEBUG:   ParentPath = [special://skin/extras/video/YourTVShows.xsp]
13:03:02.429 T:138549957872   DEBUG: Saving fileitems [special://skin/extras/video/YourTVShows.xsp]`

11 Sept -

`13:03:14.817 T:138296648944   DEBUG:   ParentPath = [special://skin/extras/video/YourTVShows.xsp]
13:03:16.374 T:138296648944   DEBUG: Saving fileitems [special://skin/extras/video/YourTVShows.xsp]`

And that was a quick one - sometimes it seemed to hang altogether and needed a restart.

Please let me know if there's anything else I can provide to help.

File: xbmc/dialogs/GUIDialogMediaFilter.cpp (hunk 2 above)
Comment (FernetMenta) [reply xbmc/dialogs/GUIDialogMediaFilter.cpp]: this is definitely slower as it was before. it creates a copy that is not used if mediaType does not match.

there might be more of those little things we missed on review

Comment (Rechi) [reply xbmc/dialogs/GUIDialogMediaFilter.cpp]: You are right, I thought I can smart and avoid to iterate over a reference as the variable is copied a few lines below anyway.

**Pull request merged**